src/Microsoft.Health.Fhir.Shared.Api/Controllers/ImportController.cs

```diff
@@ -79,9 +93,80 @@ public class ImportController : Controller
+        {
+            var sw = Stopwatch.StartNew();
+            var startDate = DateTime.UtcNow;
+            var resources = new List<ImportResource>();
+            var keys = new HashSet<ResourceKey>();
+            var fhirParser = new FhirJsonParser();
+            var importParser = new ImportResourceParser(fhirParser, _resourceWrapperFactory);
+            var index = 0L;
+            using var reader = new StreamReader(Request.Body, Encoding.UTF8);
+            try
+            {
+                if (Request.ContentType != null && Request.ContentType.Contains("application/fhir+ndjson", StringComparison.OrdinalIgnoreCase))
+                {
+                    var line = await reader.ReadLineAsync();
+                    while (line != null)
+                    {
+                        var importResource = importParser.Parse(index, 0, 0, line, ImportMode.IncrementalLoad);
+                        var key = importResource.ResourceWrapper.ToResourceKey(true);
+                        if (!keys.Add(key))
+                        {
+                            throw new RequestNotValidException(string.Format(Resources.ResourcesMustBeUnique, key));
+                        }
+
+                        resources.Add(importResource);
+                        line = await reader.ReadLineAsync();
+                        index++;
+                    }
+                }
+                else
+                {
+                    var str = await reader.ReadToEndAsync();
+                    var resource = await fhirParser.ParseAsync(str);
+                    var bundle = resource.ToResourceElement().ToPoco<Bundle>();
+                    foreach (var entry in bundle.Entry)
+                    {
+                        if (entry.Request?.Method != Bundle.HTTPVerb.PUT)
+                        {
+                            throw new RequestNotValidException(string.Format(Resources.InvalidBundleEntry, entry.Request?.Method, KnownRoutes.ImportBundle));
+                        }
+
+                        var importResource = importParser.Parse(index, 0, 0, entry.Resource, ImportMode.IncrementalLoad);
+                        var key = importResource.ResourceWrapper.ToResourceKey(true);
+                        if (!keys.Add(key))
+                        {
+                            throw new RequestNotValidException(string.Format(Resources.ResourcesMustBeUnique, key));
+                        }
+
+                        resources.Add(importResource);
+                        index++;
+                    }
+                }
+            }
+            catch (Exception)
+            {
+                throw new RequestNotValidException(string.Format(Resources.ParsingError, $"Unable to parse resource at index={index}"));
+            }
+
+            var request = new ImportBundleRequest(resources);
+            var response = await _mediator.ImportBundleAsync(request.Resources, HttpContext.RequestAborted);
```

Comment:
How are we handling exceptions here? If a SQL exception is thrown, are we exposing the exception and all details on it?

Reply:
I am using default logic. I don't think we expose details to the client in it.